### PR TITLE
feat(contracts): implement ShadowGroup multisig

### PR DESCRIPTION
**Pull request by 0xpanoramix**
## Description

Please provide a detailed description of what was done in this PR.

## Changes include

- [ ] Bugfix (non-breaking change that solves an issue).
- [ ] Hotfix (change that solves an urgent issue, and requires immediate attention).
- [ ] New feature (non-breaking change that adds functionality).
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality).
- [ ] Documentation update (change added to any kind of documentation).

## Breaking changes

Please complete this section if any breaking changes have been made, otherwise delete it.

## Checklist

- [ ] I have assigned this PR to myself.
- [ ] I have added at least 1 reviewer.
- [ ] I have added the relevant labels.
- [ ] I have updated the official documentation.
- [ ] I have added sufficient documentation in code.

## Testing

- [ ] I have tested this code with the official test suite.
- [ ] I have tested this code manually.

### Manual tests

Please complete this section if you ran manual tests for this functionality, otherwise delete it.

## Documentation update

Please link the documentation update PR in this section if it's present, otherwise delete it.

## Additional comments

Please post additional comments in this section if you have them, otherwise delete it.

### CODE REVIEW

**Review by 0xtekgrinder — CHANGES_REQUESTED** [review]
Overall, use custom errors and a few changes to do. I don't know if it is possible with semaphores but it would be better to achieve to works without transactions array and instead use the signatures and all the data of the tx in executeTransaction

File: contracts/contracts/ShadowGroup.sol [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+
+contract ShadowGroup {
+    ISemaphore public semaphore;
+    uint256 public groupID;
+
+    uint256 public quorum;
+
+    struct Transaction {
+        address to;
+        uint value;
```
[the remark this answers is not in the document]
Comment (0xpanoramix) [reply contracts/contracts/ShadowGroup.sol]: Fixed in 531d38c030abe585a8fffed5055229a54cfeabe6

File: contracts/contracts/ShadowGroup.sol [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+        );
+        require(
+            _quorum > 0 && _quorum <= _ownersIdentityCommitments.length,
+            "invalid quorum value"
+        );
+
+        semaphore = ISemaphore(_semaphoreAddress);
+        groupID = _groupID;
+
+        semaphore.createGroup(groupID, 20, address(this));
```
[the remark this answers is not in the document]
Comment (0xpanoramix) [reply contracts/contracts/ShadowGroup.sol:50]: It allows to create a group with 2^20 members and is the default value used in the Semaphore TS client.
More than enough, and can not be less than 16.

File: contracts/contracts/ShadowGroup.sol [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+        uint value;
+        bytes data;
+        bool executed;
+        uint numConfirmations;
+    }
+
+    Transaction[] public transactions;
+
+    modifier txExists(uint _txIndex) {
+        require(_txIndex < transactions.length, "tx does not exist");
```
[the remark this answers is not in the document]
Comment (0xpanoramix) [reply contracts/contracts/ShadowGroup.sol]: Fixed in 1748944f78c0ecb3d32de587ddd8fbfc6839db50.

File: contracts/contracts/ShadowGroup.sol [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+        _;
+    }
+
+    modifier notExecuted(uint _txIndex) {
+        require(!transactions[_txIndex].executed, "tx already executed");
```
[the remark this answers is not in the document]
Comment (0xpanoramix) [reply contracts/contracts/ShadowGroup.sol]: Fixed in 1748944f78c0ecb3d32de587ddd8fbfc6839db50.

File: contracts/contracts/ShadowGroup.sol [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+        transaction.executed = true;
+
+        (bool success, ) = transaction.to.call{value: transaction.value}(
+            transaction.data
+        );
+        require(success, "tx failed");
+    }
+
+    receive() external payable {
+    }
```
[the remark this answers is not in the document]
Comment (0xpanoramix) [reply contracts/contracts/ShadowGroup.sol]: Updated in e15f8dda03c973b91104f1a4e146e5d0941f74d8

File: contracts/contracts/ShadowGroup.sol [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+        uint256 _externalNullifier,
+        uint256[8] calldata _proof
+    ) txExists(_txIndex) notExecuted(_txIndex) public {
+        require(transactions[_txIndex].numConfirmations > 0, "tx has no confirmations yet");
+
+        uint256 signal = uint256(keccak256(abi.encodePacked(false)));
+        semaphore.verifyProof(groupID, _merkleTreeRoot, signal, _nullifierHash, _externalNullifier, _proof);
+
+        transactions[_txIndex].numConfirmations -= 1;
+    }
```
[the remark this answers is not in the document]
Comment (0xpanoramix) [reply contracts/contracts/ShadowGroup.sol]: Nice comment ! Updated in 95c76dc4fbf3bc28139d48133ee5f763b5dea049.

File: contracts/contracts/ShadowGroup.sol [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+
+        Transaction storage transaction = transactions[_txIndex];
+        if (transaction.numConfirmations < quorum) revert QuorumNotReached();
+
+        transaction.executed = true;
+
+        (bool success, ) = transaction.to.call{value: transaction.value}(
+            transaction.data
+        );
+        require(success, "tx failed");
```
Comment (0xtekgrinder) [inline contracts/contracts/ShadowGroup.sol:129]: custom error

**Review by 0xtekgrinder — CHANGES_REQUESTED** [review]

File: contracts/contracts/ShadowGroup.sol (hunk 7 above)
Comment (0xpanoramix) [reply contracts/contracts/ShadowGroup.sol]: Fixed in e839f2d2a1325c87a4758939063af8a1e7932ac8.

**Pull request merged**